experimental/dds/tree2/src/feature-libraries/sequence-field/compose.ts

```diff
@@ -940,15 +609,68 @@ function getIntention(
+function compareCellPositions(
+	baseCellId: DetachEvent,
+	baseMark: ExistingCellMark<unknown>,
+	newMark: EmptyInputCellMark<unknown>,
+	newRevision: RevisionTag | undefined,
+	cancelledInserts: Set<RevisionTag>,
+	gapTracker: GapTracker,
+): number {
+	const newId = getCellId(newMark, newRevision);
+	if (baseCellId.revision === newId?.revision) {
+		if (isNewAttach(newMark)) {
+			// There is some change foo that is being cancelled out as part of a rebase sandwich.
+			// The marks that make up this change (and its inverse) may be broken up differently between the base
+			// changeset and the new changeset because either changeset may have been composed with other changes
+			// whose marks may now be interleaved with the marks that represent foo/its inverse.
+			// This means that the base and new marks may not be of the same length.
+			// We do however know that the all of the marks for foo will appear in the base changeset and all of the
+			// marks for the inverse of foo will appear in the new changeset, so we can be confident that whenever
+			// we encounter such pairs of marks, they do line up such that they describe changes to the same first
+			// cell. This means we can safely treat them as inverses of one another.
+			return 0;
+		}
+		return baseCellId.index - newId.index;
+	}
+
+	const offsetInNew = getOffsetAtRevision(newMark.lineage, baseCellId.revision);
+	if (offsetInNew !== undefined) {
+		const baseOffset = gapTracker.getOffset(baseCellId.revision);
+		return offsetInNew <= baseOffset ? Infinity : baseOffset - offsetInNew;
 	}
-	const info1 = revisionMetadata.getInfo(rev1);
-	const info2 = revisionMetadata.getInfo(rev2);
-	return rev1 === info2.rollbackOf || rev2 === info1.rollbackOf;
+
+	const cmp = compareLineages(baseMark.lineage, newMark.lineage);
+	if (cmp !== 0) {
+		return Math.sign(cmp) * Infinity;
+	}
+
+	if (
+		newRevision !== undefined &&
+		newMark.type === "Insert" &&
+		cancelledInserts.has(newRevision)
+	) {
+		// We know the new insert is getting cancelled out so we need to delay returning it.
+		// The base mark that cancels the insert must appear later in the base marks.
+		return -Infinity;
+	}
+
+	if (isNewAttach(newMark)) {
+		// When the marks are at the same position, we use the tiebreak of `newMark`.
+		// TODO: Use specified tiebreak instead of always tiebreaking left.
+		return Infinity;
+	}
+
+	// `newMark` points to cells which were emptied before `baseMark` was created.
+	// We use `baseMark`'s tiebreak policy as if `newMark`'s cells were created concurrently and before `baseMark`.
+	// TODO: Use specified tiebreak instead of always tiebreaking left.
+	if (isNewAttach(baseMark)) {
+		return -Infinity;
+	}
+
+	// If `newMark`'s lineage does not overlap with `baseMark`'s,
+	// then `newMark` must be referring to cells which were created after `baseMark` was applied.
+	// The creation of those cells should happen in this composition, so they must be later in the base mark list.
```

Comment:
Elaborate. This is true because there may be any number of changesets between the base and new changesets, which the new changeset might be refering to the cells of.
TODO: add test case to exercise this.